examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,410 @@
+  PetscInt           cell_start, cell_end, dim, num_points_local, *point_cell_numbers;
+  PetscScalar       *coords_points_ref;
+  const PetscScalar *coords_points_true;
+
+  PetscFunctionBeginUser;
+
+  // Create vector to hold reference coordinates
+  {
+    Vec true_coords;
+
+    PetscCall(DMSwarmCreateLocalVectorFromField(dm_swarm, "DMSwarmPIC_coor", &true_coords));
+    PetscCall(VecDuplicate(true_coords, ref_coords));
+    PetscCall(DMSwarmDestroyLocalVectorFromField(dm_swarm, "DMSwarmPIC_coor", &true_coords));
+  }
+
+  // Allocate index set array
+  PetscCall(VecGetLocalSize(*ref_coords, &num_points_local));
+  PetscCall(DMGetDimension(dm_mesh, &dim));
+  num_points_local /= dim;
+  PetscCall(PetscMalloc1(num_points_local, &point_cell_numbers));
+
+  // Get reference coordinates for each swarm point wrt the elements in the background mesh
+  PetscCall(DMSwarmSortGetAccess(dm_swarm));
+  PetscCall(DMPlexGetHeightStratum(dm_mesh, 0, &cell_start, &cell_end));
+  PetscCall(DMSwarmGetField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coords_points_true));
+  PetscCall(VecGetArray(*ref_coords, &coords_points_ref));
+  for (PetscInt cell = cell_start; cell < cell_end; cell++) {
+    PetscReal *coords_cell_points_true, *coords_cell_points_ref;
+    PetscInt  *cell_points;
+    PetscInt   num_points_in_cell;
+
+    PetscCall(DMSwarmSortGetPointsPerCell(dm_swarm, cell, &num_points_in_cell, &cell_points));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_true));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_ref));
+    // -- Reference coordinates for swarm points in background mesh element
+    for (PetscInt p = 0; p < num_points_in_cell; p++) {
+      point_cell_numbers[cell_points[p]] = cell;
+      for (PetscInt d = 0; d < dim; d++) coords_cell_points_true[p * dim + d] = coords_points_true[cell_points[p] * dim + d];
+    }
+    PetscCall(DMPlexCoordinatesToReference(dm_mesh, cell, num_points_in_cell, coords_cell_points_true, coords_cell_points_ref));
+    for (PetscInt p = 0; p < num_points_in_cell; p++) {
+      for (PetscInt d = 0; d < dim; d++) coords_points_ref[cell_points[p] * dim + d] = coords_cell_points_ref[p * dim + d];
+    }
+
+    // -- Cleanup
+    PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_true));
+    PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_ref));
+    PetscCall(PetscFree(cell_points));
+  }
+
+  // Cleanup
+  PetscCall(DMSwarmRestoreField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coords_points_ref));
+  PetscCall(VecRestoreArrayRead(*ref_coords, &coords_points_true));
+  PetscCall(DMSwarmSortRestoreAccess(dm_swarm));
+
+  // Create index set
+  PetscCall(ISCreateGeneral(PETSC_COMM_SELF, num_points_local, point_cell_numbers, PETSC_OWN_POINTER, is_points));
```

Comment:
It is my understanding that using `PETSC_COMM_SELF` here means this will be local to the process?